File: binary_search_tree/Python/binary_search_tree.py
class Leaf:
    """
    Class that represent a leaf node of a binary tree.
    """

    def __init__(self, value=None, leftChild=None, rightChild=None):
        self.value = value
        self.leftChild = leftChild
        self.rightChild = rightChild

    def __str__(self):
        return str(self.value)

    def __repr__(self):
        return self.__str__()

    def __del__(self):
        del self.value
        del self.leftChild
        del self.rightChild

    def getChildren(self):
        return [self.leftChild, self.rightChild]


class BST:
    """
    Class that represents a Binary Search Tree.
    <https://en.wikipedia.org/wiki/Binary_search_tree>
    """

    def __init__(self):
        self.root = None

    def __del__(self):
        del self.root

    def find(self, value):
        return self.__findNode(self.root, value)

    def __findNode(self, currentNode, value):
        if currentNode is None:
            raise ValueError("Value not found in BST")
        elif value == currentNode.value:
            return currentNode
        elif value < currentNode.value:
            return self.__findNode(currentNode.leftChild, value)
        else:
            return self.__findNode(currentNode.rightChild, value)

    def insert(self, value):
        self.root = self.__insertNode(self.root, value)

    def __insertNode(self, currentNode, value):
        if currentNode is None:
            return Leaf(value)
        elif value < currentNode.value:
            currentNode.leftChild = self.__insertNode(currentNode.leftChild,
                                                      value)
        elif value > currentNode.value:
            currentNode.rightChild = self.__insertNode(currentNode.rightChild,
                                                       value)
        else:
            currentNode.value = value
        return currentNode


def main():
    values = [1, 2, 0, 1.5, 0.5]
    tree = BST()
    for n in values:
        tree.insert(n)

    print(tree.find(0).rightChild)  # 0.5
    print(tree.root.rightChild)  # 2


if __name__ == '__main__':
    main()

File: binary_search_tree/java/binary_search_tree.java
class binary_search_tree {
//Class defining the node
	class Node {
		int key;
		Node left, right;
		public Node(int item) {
			key = item;
			left = right = null;
		}
	}
	// Root of Binary Search Tree
	Node root;
	// Constructor initiates BST root as null
	binary_search_tree() {
		root = null;
	}
	// This method mainly calls insertNode()
	void insert(int key) {
	root = insertNode(root, key);
	}
	/* A recursive function to insert a new key in BST */
	Node insertNode(Node root, int key) {

		/* If the tree is empty, return a new node */
		if (root == null) {
			root = new Node(key);
			return root;
		}

		/* Otherwise, recursively move down the tree */
		if (key < root.key)
			root.left = insertNode(root.left, key);
		else if (key > root.key)
			root.right = insertNode(root.right, key);

		/* return the (unchanged) node pointer */
		return root;
	}
	// This method mainly calls InorderNode()
	void inorder() {
	inorderNode(root);
	}

	// A utility function to do inorder traversal of BST
	void inorderNode(Node root) {
		if (root != null) {
			inorderNode(root.left);
			System.out.println(root.key);
			inorderNode(root.right);
		}
	}
	// Driver Program to test above functions
	public static void main(String[] args) {
		binary_search_tree tree = new binary_search_tree();
		tree.insert(50);
		tree.insert(30);
		tree.insert(20);
		tree.insert(40);
		tree.insert(70);
		tree.insert(60);
		tree.insert(80);
		// print inorder traversal of the Binary Search Tree
		tree.inorder();
	}
}
